### PR TITLE
Feat/EditorCommands

### PR DESCRIPTION
**Pull request by maimonav**
- EditorCommands

### CODE REVIEW

**Comment by oded-soffrin** [comment]
@shaulgo can we prioritize this PR review?

**Comment by oded-soffrin** [comment]
@shaulgo ?

File: packages/editor/web/src/RichContentEditor/EditorCommands.ts [hunk 1]
```diff
@@ -0,0 +1,250 @@
...
+    setTextAlignment: (textAlignment: TextAlignment): void =>
+      setEditorState(setTextAlignment(getEditorState(), textAlignment)),
+    setSelection,
+  };
+
+  const pluginsCommands = {
+    insertBlock: <K extends keyof PluginsDataMap>(
+      type: K,
+      data?: PluginsDataMap[K],
+      settings?: { forceSelection?: boolean; isRicosSchema?: boolean }
```
Comment (shaulgo) [inline packages/editor/web/src/RichContentEditor/EditorCommands.ts:157]: Let's try to remove forceSelection.

File: packages/editor/web/src/RichContentEditor/EditorCommands.ts [hunk 2]
```diff
@@ -0,0 +1,250 @@
...
+    updateBlock: <K extends keyof PluginsDataMap>(
+      blockKey: string,
+      type: K,
+      data?: PluginsDataMap[K],
+      settings?: { forceSelection?: boolean; isRicosSchema?: boolean; useCurrentData?: boolean }
+    ) => {
+      const { forceSelection = true } = settings || {};
+      const draftType = PLUGIN_TYPE_MAP[type];
+      const { [draftType]: createPluginData } = createPluginsDataMap;
+      if (createPluginData) {
```
Comment (shaulgo) [inline packages/editor/web/src/RichContentEditor/EditorCommands.ts:187]: no need for this

**Review by shaulgo — CHANGES_REQUESTED** [review]

**Review by shaulgo — CHANGES_REQUESTED** [review]

File: packages/editor/web/src/RichContentEditor/EditorCommands.ts [hunk 3]
```diff
@@ -0,0 +1,250 @@
...
+    insertDecoration: <K extends keyof DecorationsDataMap>(
+      type: K,
+      data?: DecorationsDataMap[K],
+      settings?: { isRicosSchema?: boolean }
+    ) => {
+      const draftType = PLUGIN_TYPE_MAP[type];
+      const { [draftType]: createPluginData } = createPluginsDataMap;
+      if (createPluginData) {
+        const pluginData = createPluginData(data, undefined, settings?.isRicosSchema);
+        if (pluginData) {
```
[the remark this answers is not in the document]
Comment (maimonav) [reply packages/editor/web/src/RichContentEditor/EditorCommands.ts]: Done.

File: packages/editor/web/src/RichContentEditor/EditorCommands.ts [hunk 4]
```diff
@@ -0,0 +1,250 @@
...
+
+  const decorationsCommands = {
+    insertDecoration: <K extends keyof DecorationsDataMap>(
+      type: K,
+      data?: DecorationsDataMap[K],
+      settings?: { isRicosSchema?: boolean }
+    ) => {
+      const draftType = PLUGIN_TYPE_MAP[type];
+      const { [draftType]: createPluginData } = createPluginsDataMap;
+      if (createPluginData) {
```
[the remark this answers is not in the document]
Comment (maimonav) [reply packages/editor/web/src/RichContentEditor/EditorCommands.ts]: Done.

File: packages/editor/web/src/RichContentEditor/EditorCommands.ts [hunk 5]
```diff
@@ -0,0 +1,250 @@
...
+          );
+          setEditorState(
+            forceSelection
+              ? EditorState.forceSelection(newEditorState, newSelection)
+              : EditorState.acceptSelection(newEditorState, newSelection)
+          );
+        }
+      }
+    },
+    updateBlock: <K extends keyof PluginsDataMap>(
```
[the remark this answers is not in the document]
Comment (maimonav) [reply packages/editor/web/src/RichContentEditor/EditorCommands.ts]: Done.

File: packages/editor/web/src/RichContentEditor/EditorCommands.ts (hunk 2 above)
Comment (maimonav) [reply packages/editor/web/src/RichContentEditor/EditorCommands.ts]: Done.

File: packages/editor/web/src/RichContentEditor/EditorCommands.ts (hunk 1 above)
Comment (maimonav) [reply packages/editor/web/src/RichContentEditor/EditorCommands.ts]: Done.

File: packages/editor/web/src/RichContentEditor/EditorCommands.ts [hunk 6]
```diff
@@ -0,0 +1,250 @@
...
+        getEditorState(),
+        SelectionState.createEmpty(blockKey).merge(selection || {})
+      )
+    );
+
+  const editorState = {
+    // TODO: check if needed, plus type error using SelectionState, not sure why
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    getSelection: (): any => getEditorState().getSelection(),
+    getSelectedBlockKey: () =>
```
[the remark this answers is not in the document]
Comment (maimonav) [reply packages/editor/web/src/RichContentEditor/EditorCommands.ts]: Done.

File: packages/common/web/src/types/pluginTypes.ts [hunk 7]
```diff
@@ -53,13 +59,89 @@ import {
...
+} from 'ricos-schema';
+export { Node_Type, Decoration_Type, LinkData };
+
+export type CreatePluginData<PluginData> = (
+  pluginData?: PluginData,
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  currentData?: any,
+  isRicosSchema?: boolean
+) => // eslint-disable-next-line @typescript-eslint/no-explicit-any
+Record<string, any> | undefined;
```
[the remark this answers is not in the document]
Comment (maimonav) [reply packages/common/web/src/types/pluginTypes.ts]: Removed.

File: packages/ricos-editor/web/src/RicosEditor.spec.tsx [hunk 8]
```diff
@@ -45,6 +81,83 @@ const getRCE = (ricosEditorProps?: RicosEditorProps, asWrapper?: boolean) => {
...
+  pluginName,
+  { type, data, newData, expectedNewData },
+]) =>
+  !isMention(type) &&
+  it(`should set ${pluginName}`, () => {
+    const ricosEditor = getRicosEditorInstance({ plugins, content }) as RicosEditor;
+    ricosEditor.getEditorCommands().setSelection(blockKey, selection);
+    ricosEditor.getEditorCommands().insertDecoration(type, data, settings);
+    ricosEditor.getEditorCommands().setSelection(blockKey, selection);
+    ricosEditor.getEditorCommands().insertDecoration(type, newData, settings);
```
Comment (shaulgo) [inline packages/ricos-editor/web/src/RicosEditor.spec.tsx:145]: Does this change the whole entity?

File: packages/ricos-editor/web/src/RicosEditor.spec.tsx [hunk 9]
```diff
@@ -163,4 +276,107 @@ describe('RicosEditor', () => {
       expect({ openModal, closeModal }).toStrictEqual(modalSettings);
     });
   });
+  describe('Editor Commands API', () => {
+    describe('Editor text formatting API', () => {
+      it('should have left text alignment', () => {
+        const ricosEditor = getRicosEditorInstance({ content }) as RicosEditor;
+        ricosEditor.getEditorCommands().setSelection('o12', {});
```
Comment (shaulgo) [inline packages/ricos-editor/web/src/RicosEditor.spec.tsx:283]: o12?

File: packages/ricos-editor/web/src/RicosEditor.spec.tsx [hunk 10]
```diff
@@ -163,4 +276,107 @@ describe('RicosEditor', () => {
...
+        ricosEditor.getEditorCommands().toggleInlineStyle('italic');
+        expect(ricosEditor.getEditorCommands().hasInlineStyle('italic')).toBeTruthy();
+      });
+      it('should not have italic inline style', async () => {
+        const ricosEditor = getRicosEditorInstance({ content }) as RicosEditor;
+        ricosEditor.getEditorCommands().setSelection(blockKey, selection);
+        ricosEditor.getEditorCommands().toggleInlineStyle('italic');
+        ricosEditor.getEditorCommands().toggleInlineStyle('italic');
+        expect(ricosEditor.getEditorCommands().hasInlineStyle('italic')).toBeFalsy();
+      });
```
Comment (shaulgo) [inline packages/ricos-editor/web/src/RicosEditor.spec.tsx:319]: let's make generic functions
testInlineStyle('bold')
testInlineStyle('underline')
...

**Review by shaulgo — CHANGES_REQUESTED** [review]

File: packages/editor-common/web/src/Utils/draftUtils.ts [hunk 11]
```diff
@@ -308,6 +321,12 @@ export const getAnchorBlockData = (editorState: EditorState) => {
   return block.get('data').toJS();
 };
 
+export const updateEntityData = (editorState: EditorState, blockKey: string, data) => {
```
[the remark this answers is not in the document]
Comment (maimonav) [reply packages/editor-common/web/src/Utils/draftUtils.ts:324]: Removed updateEntityData, added `blockKeyToEntityKey`.

File: packages/plugin-commons/web/src/Base/toolbars/createAtomicPluginToolbar.jsx [hunk 12]
```diff
@@ -134,7 +134,7 @@ export default function createAtomicPluginToolbar({
       if (!this.shouldCreate) {
         return;
       }
-      if (focusedBlock && !isSSR() && window.getSelection().isCollapsed) {
+      if (focusedBlock && !isSSR() && window.getSelection?.().isCollapsed) {
```
[the remark this answers is not in the document]
Comment (maimonav) [reply packages/plugin-commons/web/src/Base/toolbars/createAtomicPluginToolbar.jsx:137]: Done

File: packages/ricos-editor/web/src/RicosEditor.spec.tsx [hunk 13]
```diff
@@ -1,17 +1,53 @@
...
+  selection,
+  selectionCollapsed,
+  endOfSelection,
+  mentionSelection,
+  pluginsTestConfig,
+  decorationsTestConfig,
+} from './utils/editorCommandsUtil';
 import Enzyme from 'enzyme';
 import Adapter from 'enzyme-adapter-react-16';
+import 'mutationobserver-shim';
```
[the remark this answers is not in the document]
Comment (maimonav) [reply packages/ricos-editor/web/src/RicosEditor.spec.tsx:32]: Done

File: packages/ricos-editor/web/src/RicosEditor.spec.tsx [hunk 14]
```diff
@@ -45,6 +81,83 @@ const getRCE = (ricosEditorProps?: RicosEditorProps, asWrapper?: boolean) => {
   return element.at(element.length - 1); // due to add html by strategies
 };
 
+type Settings = { isRicosSchema?: boolean; forceSelection?: boolean };
```
[the remark this answers is not in the document]
Comment (maimonav) [reply packages/ricos-editor/web/src/RicosEditor.spec.tsx:84]: Done

File: package.json [hunk 15]
```diff
@@ -152,6 +152,7 @@
 		"lerna": "^3.21.0",
 		"lodash": "^4.17.15",
 		"mini-css-extract-plugin": "^0.11.3",
+		"mutationobserver-shim": "^0.3.7",
```
[the remark this answers is not in the document]
Comment (maimonav) [reply package.json:156]: Removed.

File: packages/editor/web/src/RichContentEditor/EditorCommands.ts [hunk 16]
```diff
@@ -0,0 +1,250 @@
...
+  RICOS_VIDEO_TYPE,
+  RICOS_POLL_TYPE,
+  RICOS_FILE_TYPE,
+  RICOS_LINK_TYPE,
+  LINK_TYPE,
+  RICOS_MENTION_TYPE,
+  MENTION_TYPE,
+} from 'wix-rich-content-common';
+
+type TextBlockType =
```
[the remark this answers is not in the document]
Comment (maimonav) [reply packages/editor/web/src/RichContentEditor/EditorCommands.ts:51]: Done

File: packages/editor/web/src/RichContentEditor/EditorCommands.ts [hunk 17]
```diff
@@ -0,0 +1,250 @@
...
+    isRedoStackEmpty: () => getEditorState().getRedoStack().size === 0,
+    hasLinkInSelection: () => hasLinksInSelection(getEditorState()),
+    getLinkDataInSelection: () => getLinkDataInSelection(getEditorState()),
+    getSelectedBlockData: () => getEntityData(getEditorState()) || {},
+  };
+
+  const textFormattingCommands = {
+    undo: (): void => setEditorState(undo(getEditorState())),
+    redo: (): void => setEditorState(redo(getEditorState())),
+    toggleInlineStyle: (style: InlineStyle): void =>
```
[the remark this answers is not in the document]
Comment (maimonav) [reply packages/editor/web/src/RichContentEditor/EditorCommands.ts:141]: Tried it, it makes the consumption harder.
User will need to import enum InlineStyle and Do `InlineStyle.Bold` for example.

File: packages/editor/web/src/RichContentEditor/EditorCommands.ts [hunk 18]
```diff
@@ -0,0 +1,250 @@
...
+    triggerDecoration: <K extends keyof Omit<DecorationsDataMap, typeof RICOS_LINK_TYPE>>(
+      type: K
+    ) => {
+      const newEditorState = triggerDecorationsMap[type]?.(getEditorState());
+      if (newEditorState) {
+        setEditorState(newEditorState);
+      }
+    },
+    deleteDecoration: <K extends keyof Omit<DecorationsDataMap, typeof RICOS_MENTION_TYPE>>(
+      type: K
```
[the remark this answers is not in the document]
Comment (maimonav) [reply packages/editor/web/src/RichContentEditor/EditorCommands.ts:205]: Yes

File: packages/ricos-editor/web/src/utils/editorCommandsUtil.ts [hunk 19]
```diff
@@ -0,0 +1,538 @@
...
+
+const divider = {
+  type: RICOS_DIVIDER_TYPE,
+  nodeType: Node_Type.DIVIDER,
+  data: DividerData.fromJSON({
+    type: DividerData_DividerType.DOUBLE,
+  }),
+  newData: DividerData.fromJSON({
+    type: DividerData_DividerType.DASHED,
+  }),
```
[the remark this answers is not in the document]
Comment (maimonav) [reply packages/ricos-editor/web/src/utils/editorCommandsUtil.ts:85]: Done.

File: packages/ricos-editor/web/src/utils/editorCommandsUtil.ts [hunk 20]
```diff
@@ -0,0 +1,538 @@
...
+  expectedData: {
+    type: 'double',
+    config: { size: 'large', alignment: 'center', textWrap: 'nowrap' },
+  },
+  expectedNewData: {
+    type: 'dashed',
+    config: { size: 'large', alignment: 'center', textWrap: 'nowrap' },
+  },
```
[the remark this answers is not in the document]
Comment (maimonav) [reply packages/ricos-editor/web/src/utils/editorCommandsUtil.ts:93]: Done

File: packages/ricos-editor/web/src/RicosEditor.spec.tsx [hunk 21]
```diff
@@ -45,6 +81,83 @@ const getRCE = (ricosEditorProps?: RicosEditorProps, asWrapper?: boolean) => {
...
+    expect(ricosEditor.getEditorCommands().getSelectedBlockData()).toEqual(expectedNewData);
+  });
+
+const deletePluginTest = (settings: Settings) => ([pluginName, { type, nodeType, data }]) =>
+  it(`should remove ${pluginName}`, () => {
+    const ricosEditor = getRicosEditorInstance({ plugins, content }) as RicosEditor;
+    data = settings?.isRicosSchema ? data : convertNodeDataToDraft(nodeType, data);
+    const blockKey = ricosEditor.getEditorCommands().insertBlock(type, data, settings);
+    ricosEditor.getEditorCommands().deleteBlock(blockKey);
+    expect(ricosEditor.getEditorCommands().getSelectedBlockData()).toEqual({});
```
[the remark this answers is not in the document]
Comment (maimonav) [reply packages/ricos-editor/web/src/RicosEditor.spec.tsx:118]: Done.

File: packages/ricos-editor/web/src/RicosEditor.spec.tsx [hunk 22]
```diff
@@ -45,6 +81,83 @@ const getRCE = (ricosEditorProps?: RicosEditorProps, asWrapper?: boolean) => {
...
+  });
+
+const insertDecorationTest = (settings: Settings) => ([pluginName, { type, data, expectedData }]) =>
+  it(`should insert ${pluginName}`, () => {
+    const ricosEditor = getRicosEditorInstance({ plugins, content }) as RicosEditor;
+    ricosEditor
+      .getEditorCommands()
+      .setSelection(blockKey, isMention(type) ? endOfSelection : selection);
```
[the remark this answers is not in the document]
Comment (maimonav) [reply packages/ricos-editor/web/src/RicosEditor.spec.tsx:128]: Done.

File: packages/ricos-editor/web/src/RicosEditor.spec.tsx (hunk 9 above)
Comment (maimonav) [reply packages/ricos-editor/web/src/RicosEditor.spec.tsx:299]: Done.

File: packages/ricos-editor/web/src/RicosEditor.spec.tsx (hunk 10 above)
Comment (maimonav) [reply packages/ricos-editor/web/src/RicosEditor.spec.tsx:335]: Done.

File: packages/ricos-editor/web/src/RicosEditor.spec.tsx (hunk 8 above)
Comment (maimonav) [reply packages/ricos-editor/web/src/RicosEditor.spec.tsx]: It works like it should.
-  if you select partial selection of block - it adds a link on it.
- if you select all block and it contains other links - it overrides them and change the block entity to single link for all block.

File: packages/plugin-link/web/src/createLinkData.ts [hunk 23]
```diff
@@ -0,0 +1,21 @@
...
+    linkData = { ...linkData, nofollow: true };
+  } else {
+    linkData = { ...linkData, relValue: data.rel || 'noopener' };
+  }
+  return { ...linkData, url: data.url, href: data.href };
+};
+
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+export const createLinkData: CreatePluginsDataMap[typeof RICOS_LINK_TYPE] = pluginData => {
+  return pluginData ? convertLinkData(pluginData) : {};
```
[the remark this answers is not in the document]
Comment (maimonav) [reply packages/plugin-link/web/src/createLinkData.ts:20]: Done

**Pull request merged**